File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/board/screen/BoardWriteViewModel.kt
package com.ssafy.ssabree.features.board.screen

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import android.util.Log
import androidx.exifinterface.media.ExifInterface
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.BoardRepository
import com.ssafy.ssabree.core.repository.PostRepository
import com.ssafy.ssabree.core.repository.UploadRepository
import com.ssafy.ssabree.core.repository.model.PollCreateInfo
import com.ssafy.ssabree.core.repository.model.PostCreateInfo
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.io.File
import kotlin.math.max
import kotlin.math.roundToInt

private const val TAG = "BoardWriteViewModel"

data class VoteOption(
    val id: Int,
    val text: String,
)

data class BoardOption(
    val id: Long,
    val name: String
)

data class BoardWriteUiState(
    val title: String = "",
    val content: String = "",
    val attachedImages: List<Uri> = emptyList(),
    val boardOptions: List<BoardOption> = emptyList(),
    val selectedBoardId: Long? = null,
    val isBoardMenuExpanded: Boolean = false,
    val isVoteEnabled: Boolean = false,
    val voteTitle: String = "",
    val voteOptions: List<VoteOption> = listOf(
        VoteOption(id = 0, text = ""),
        VoteOption(id = 1, text = ""),
    ),
    val isSubmitting: Boolean = false,
    val isSubmitSuccess: Boolean = false,
    val submitError: String? = null,
) {
    val isSubmitEnabled: Boolean
        get() = title.isNotBlank() &&
            content.isNotBlank() &&
            selectedBoardId != null &&
            isPollValid &&
            !isSubmitting

    val hasContent: Boolean
        get() = title.isNotBlank() || content.isNotBlank() || attachedImages.isNotEmpty()

    val selectedBoardName: String
        get() = boardOptions.firstOrNull { it.id == selectedBoardId }?.name ?: "게시판 선택"

    private val isPollValid: Boolean
        get() = !isVoteEnabled || (voteTitle.isNotBlank() && voteOptions.count { it.text.isNotBlank() } >= 2)
}

class BoardWriteViewModel(
    private val postRepository: PostRepository,
    private val boardRepository: BoardRepository,
    private val uploadRepository: UploadRepository,
    private val appContext: Context
) : ViewModel() {

    companion object {
        private const val MAX_IMAGE_DIMENSION = 1280
        private const val MAX_IMAGE_BYTES = 1_000_000 // ~1MB
        private const val MAX_IMAGE_COUNT = 5
    }

    private val _uiState = MutableStateFlow(BoardWriteUiState())
    val uiState: StateFlow<BoardWriteUiState> = _uiState.asStateFlow()

    init {
        loadBoards()
    }

    fun onTitleChange(value: String) {
        _uiState.update { it.copy(title = value) }
    }

    fun onContentChange(value: String) {
        _uiState.update { it.copy(content = value) }
    }

    fun onImagesAttached(uris: List<Uri>) {
        if (uris.isEmpty()) return
        _uiState.update { state ->
            val remaining = MAX_IMAGE_COUNT - state.attachedImages.size
            if (remaining <= 0) return@update state
            val toAdd = uris.take(remaining)
            state.copy(attachedImages = state.attachedImages + toAdd)
        }
    }

    fun onRemoveImage(uri: Uri) {
        _uiState.update { state ->
            state.copy(attachedImages = state.attachedImages - uri)
        }
    }

    fun onToggleBoardMenu(expanded: Boolean) {
        _uiState.update { it.copy(isBoardMenuExpanded = expanded) }
    }

    fun onBoardSelected(boardId: Long) {
        _uiState.update {
            it.copy(
                selectedBoardId = boardId,
                isBoardMenuExpanded = false
            )
        }
    }

    fun onToggleVote() {
        _uiState.update { state ->
            state.copy(
                isVoteEnabled = !state.isVoteEnabled,
                voteTitle = if (state.isVoteEnabled) "" else state.voteTitle,
                voteOptions = if (state.isVoteEnabled) listOf(
                    VoteOption(id = 0, text = ""),
                    VoteOption(id = 1, text = ""),
                ) else state.voteOptions,
            )
        }
    }

    fun onVoteTitleChange(value: String) {
        _uiState.update { it.copy(voteTitle = value) }
    }

    fun onVoteOptionChange(id: Int, value: String) {
        _uiState.update { state ->
            state.copy(
                voteOptions = state.voteOptions.map {
                    if (it.id == id) it.copy(text = value) else it
                }
            )
        }
    }

    fun onAddVoteOption() {
        _uiState.update { state ->
            if (state.voteOptions.size >= 5) return@update state
            val nextId = (state.voteOptions.maxOfOrNull { it.id } ?: -1) + 1
            state.copy(voteOptions = state.voteOptions + VoteOption(id = nextId, text = ""))
        }
    }

    fun onRemoveVoteOption(id: Int) {
        _uiState.update { state ->
            if (state.voteOptions.size <= 2) return@update state
            state.copy(voteOptions = state.voteOptions.filter { it.id != id })
        }
    }

    fun onSubmit() {
        if (!_uiState.value.isSubmitEnabled) return
        val selectedBoardId = _uiState.value.selectedBoardId
        if (selectedBoardId == null) {
            _uiState.update { it.copy(submitError = "게시판을 선택해주세요.") }
            return
        }

        _uiState.update { it.copy(isSubmitting = true, submitError = null) }
        viewModelScope.launch {
            val state = _uiState.value
            val imageUrlsResult = runCatching {
                if (state.attachedImages.isEmpty()) {
                    emptyList()
                } else {
                    state.attachedImages.map { uri ->
                        val file = prepareUploadFile(uri).getOrThrow()
                        uploadRepository.uploadImage(file).getOrThrow()
                    }
                }
            }

            val pollInfo = buildPollInfo(state)

            imageUrlsResult
                .onFailure { e ->
                    Log.d(TAG, "image upload failed (${e.message})")
                    _uiState.update {
                        it.copy(
                            isSubmitting = false,
                            submitError = e.message ?: "이미지 업로드에 실패했습니다."
                        )
                    }
                }
                .onSuccess { imageUrls ->
                    postRepository.createPost(
                        PostCreateInfo(
                            title = state.title,
                            content = state.content,
                            boardId = selectedBoardId,
                            imageUrls = imageUrls,
                            poll = pollInfo
                        )
                    ).onFailure { e ->
                        Log.d(TAG, "onSubmit: failed (${e.message})")
                        _uiState.update {
                            it.copy(
                                isSubmitting = false,
                                submitError = e.message ?: "알 수 없는 오류가 발생했습니다."
                            )
                        }
                    }.onSuccess {
                        Log.d(TAG, "onSubmit: succeed")
                        _uiState.update {
                            it.copy(
                                isSubmitting = false,
                                isSubmitSuccess = true
                            )
                        }
                    }
                }
        }
    }

    fun clearSubmitError() {
        _uiState.update { it.copy(submitError = null) }
    }

    fun clearSubmitSuccess() {
        _uiState.update { it.copy(isSubmitSuccess = false) }
    }

    private fun loadBoards() {
        viewModelScope.launch {
            boardRepository.getBoards()
                .onSuccess { boards ->
                    val options = boards.map { BoardOption(it.id, it.name) }
                    val defaultId = options.firstOrNull()?.id
                    _uiState.update {
                        it.copy(
                            boardOptions = options,
                            selectedBoardId = it.selectedBoardId ?: defaultId
                        )
                    }
                }
                .onFailure { e ->
                    Log.d(TAG, "loadBoards: failed (${e.message})")
                    _uiState.update {
                        it.copy(submitError = it.submitError ?: "게시판 목록을 불러오지 못했습니다.")
                    }
                }
        }
    }

    private fun buildPollInfo(state: BoardWriteUiState): PollCreateInfo? {
        if (!state.isVoteEnabled) return null
        val title = state.voteTitle.trim()
        val options = state.voteOptions.map { it.text.trim() }.filter { it.isNotBlank() }
        return if (title.isNotBlank() && options.size >= 2) {
            PollCreateInfo(title = title, options = options)
        } else {
            null
        }
    }

    private suspend fun prepareUploadFile(uri: Uri): Result<File> = withContext(Dispatchers.IO) {
        runCatching {
            val resolver = appContext.contentResolver
            val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
            resolver.openInputStream(uri)?.use { BitmapFactory.decodeStream(it, null, options) }

            options.inSampleSize = calculateInSampleSize(options, MAX_IMAGE_DIMENSION, MAX_IMAGE_DIMENSION)
            options.inJustDecodeBounds = false

            val rotation = readRotationDegrees(uri)

            val bitmap = resolver.openInputStream(uri)?.use { BitmapFactory.decodeStream(it, null, options) }
                ?: throw IllegalArgumentException("이미지를 불러올 수 없습니다.")

            val rotatedBitmap = rotateBitmapIfNeeded(bitmap, rotation)
            val scaledBitmap = scaleBitmap(rotatedBitmap, MAX_IMAGE_DIMENSION)
            val compressed = compressBitmapToSize(scaledBitmap, MAX_IMAGE_BYTES)

            val outputFile = File.createTempFile("post_", ".jpg", appContext.cacheDir)
            outputFile.outputStream().use { it.write(compressed) }

            if (scaledBitmap != rotatedBitmap) scaledBitmap.recycle()
            if (rotatedBitmap != bitmap) rotatedBitmap.recycle()
            bitmap.recycle()

            outputFile
        }
    }

    private fun calculateInSampleSize(options: BitmapFactory.Options, reqWidth: Int, reqHeight: Int): Int {
        val (height: Int, width: Int) = options.run { outHeight to outWidth }
        var inSampleSize = 1

        if (height > reqHeight || width > reqWidth) {
            val halfHeight: Int = height / 2
            val halfWidth: Int = width / 2

            while ((halfHeight / inSampleSize) >= reqHeight && (halfWidth / inSampleSize) >= reqWidth) {
                inSampleSize *= 2
            }
        }
        return inSampleSize
    }

    private fun scaleBitmap(bitmap: Bitmap, maxDimension: Int): Bitmap {
        val largestSide = max(bitmap.width, bitmap.height).toFloat()
        if (largestSide <= maxDimension) return bitmap
        val scale = maxDimension / largestSide
        val targetWidth = (bitmap.width * scale).roundToInt().coerceAtLeast(1)
        val targetHeight = (bitmap.height * scale).roundToInt().coerceAtLeast(1)
        return Bitmap.createScaledBitmap(bitmap, targetWidth, targetHeight, true)
    }

    private fun compressBitmapToSize(bitmap: Bitmap, maxBytes: Int): ByteArray {
        var quality = 90
        var compressed: ByteArray
        do {
            val bos = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, quality, bos)
            compressed = bos.toByteArray()
            quality -= 10
        } while (compressed.size > maxBytes && quality > 30)
        if (compressed.size > maxBytes) {
            throw IllegalArgumentException("이미지 용량이 너무 큽니다. 1MB 이하로 줄여주세요.")
        }
        return compressed
    }

    private fun readRotationDegrees(uri: Uri): Int {
        return try {
            appContext.contentResolver.openInputStream(uri)?.use { input ->
                ExifInterface(input).let { exif ->
                    when (exif.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)) {
                        ExifInterface.ORIENTATION_ROTATE_90 -> 90
                        ExifInterface.ORIENTATION_ROTATE_180 -> 180
                        ExifInterface.ORIENTATION_ROTATE_270 -> 270
                        else -> 0
                    }
                }
            } ?: 0
        } catch (e: Exception) {
            0
        }
    }

    private fun rotateBitmapIfNeeded(bitmap: Bitmap, degrees: Int): Bitmap {
        if (degrees == 0) return bitmap
        val matrix = Matrix().apply { postRotate(degrees.toFloat()) }
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    }
}

File: iOS/ssabree/ssabree/Features/Board/BoardWriteViewModel.swift
import Foundation
import Observation
import UIKit

// MARK: - Vote Option Model

struct VoteOptionModel: Identifiable {
    let id: Int
    var text: String
}

// MARK: - Board Option Model

struct BoardOptionModel: Identifiable {
    let id: Int
    let name: String
}

// MARK: - UI State

struct BoardWriteUiState {
    var title: String = ""
    var content: String = ""
    var attachedImages: [UIImage] = []
    var boardOptions: [BoardOptionModel] = []
    var selectedBoardId: Int? = nil
    var isBoardMenuExpanded: Bool = false
    var isVoteEnabled: Bool = false
    var voteTitle: String = ""
    var voteOptions: [VoteOptionModel] = [
        VoteOptionModel(id: 0, text: ""),
        VoteOptionModel(id: 1, text: "")
    ]
    var isSubmitting: Bool = false
    var isSubmitSuccess: Bool = false
    var submitError: String? = nil

    // Computed properties
    var isSubmitEnabled: Bool {
        !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        selectedBoardId != nil &&
        isPollValid &&
        !isSubmitting
    }

    var hasContent: Bool {
        !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
        !content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
        !attachedImages.isEmpty
    }

    var selectedBoardName: String {
        boardOptions.first { $0.id == selectedBoardId }?.name ?? "게시판 선택"
    }

    private var isPollValid: Bool {
        if !isVoteEnabled { return true }
        let validTitle = !voteTitle.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        let validOptionsCount = voteOptions.filter { !$0.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }.count >= 2
        return validTitle && validOptionsCount
    }
}

// MARK: - ViewModel

@Observable
@MainActor
final class BoardWriteViewModel {
    private let postRepository: PostRepository
    private let boardRepository: BoardRepository
    private let uploadRepository: UploadRepository
    private let initialBoardId: Int?

    var uiState = BoardWriteUiState()

    init(
        postRepository: PostRepository,
        boardRepository: BoardRepository,
        uploadRepository: UploadRepository,
        initialBoardId: Int? = nil
    ) {
        self.postRepository = postRepository
        self.boardRepository = boardRepository
        self.uploadRepository = uploadRepository
        self.initialBoardId = initialBoardId

        // Pre-select board if provided
        if let boardId = initialBoardId {
            uiState.selectedBoardId = boardId
        }

        Task {
            await loadBoards()
        }
    }

    // MARK: - Title & Content

    func onTitleChange(_ value: String) {
        uiState.title = value
    }

    func onContentChange(_ value: String) {
        uiState.content = value
    }

    // MARK: - Board Selection

    func onToggleBoardMenu(_ expanded: Bool) {
        uiState.isBoardMenuExpanded = expanded
    }

    func onBoardSelected(_ boardId: Int) {
        uiState.selectedBoardId = boardId
        uiState.isBoardMenuExpanded = false
    }

    // MARK: - Image Attachment

    func onImageAttached(_ image: UIImage) {
        // 최대 5개까지 이미지 추가 가능
        guard uiState.attachedImages.count < 5 else { return }
        uiState.attachedImages.append(image)
    }

    var canAddMoreImages: Bool {
        uiState.attachedImages.count < 5
    }

    func onRemoveImage(_ image: UIImage) {
        uiState.attachedImages.removeAll { $0 === image }
    }

    func onRemoveImageAt(_ index: Int) {
        guard index >= 0 && index < uiState.attachedImages.count else { return }
        uiState.attachedImages.remove(at: index)
    }

    // MARK: - Vote

    func onToggleVote() {
        uiState.isVoteEnabled.toggle()
        if !uiState.isVoteEnabled {
            // Reset vote state when disabled
            uiState.voteTitle = ""
            uiState.voteOptions = [
                VoteOptionModel(id: 0, text: ""),
                VoteOptionModel(id: 1, text: "")
            ]
        }
    }

    func onVoteTitleChange(_ value: String) {
        uiState.voteTitle = value
    }

    func onVoteOptionChange(id: Int, text: String) {
        if let index = uiState.voteOptions.firstIndex(where: { $0.id == id }) {
            uiState.voteOptions[index].text = text
        }
    }

    func onAddVoteOption() {
        guard uiState.voteOptions.count < 5 else { return }
        let nextId = (uiState.voteOptions.map { $0.id }.max() ?? -1) + 1
        uiState.voteOptions.append(VoteOptionModel(id: nextId, text: ""))
    }

    func onRemoveVoteOption(id: Int) {
        guard uiState.voteOptions.count > 2 else { return }
        uiState.voteOptions.removeAll { $0.id == id }
    }

    // MARK: - Submit

    func onSubmit() async {
        guard uiState.isSubmitEnabled else { return }
        guard let selectedBoardId = uiState.selectedBoardId else {
            uiState.submitError = "게시판을 선택해주세요."
            return
        }

        uiState.isSubmitting = true
        uiState.submitError = nil

        // Upload images if attached
        var imageUrls: [String] = []
        for image in uiState.attachedImages {
            let uploadResult = await uploadImage(image)
            switch uploadResult {
            case .success(let url):
                imageUrls.append(url)
            case .failure(let error):
                uiState.isSubmitting = false
                uiState.submitError = "이미지 업로드에 실패했습니다. \(error.localizedDescription)"
                return
            }
        }

        // Build poll info if enabled
        let pollInfo = buildPollInfo()

        // Create post
        let postInfo = PostCreateInfo(
            title: uiState.title.trimmingCharacters(in: .whitespacesAndNewlines),
            content: uiState.content.trimmingCharacters(in: .whitespacesAndNewlines),
            boardId: selectedBoardId,
            poll: pollInfo,
            images: imageUrls
        )

        let result = await postRepository.createPost(post: postInfo)

        switch result {
        case .success:
            uiState.isSubmitSuccess = true
        case .failure(let error):
            uiState.submitError = "게시글 작성에 실패했습니다. \(error.localizedDescription)"
        }

        uiState.isSubmitting = false
    }

    func clearSubmitError() {
        uiState.submitError = nil
    }

    func clearSubmitSuccess() {
        uiState.isSubmitSuccess = false
    }

    // MARK: - Private Methods

    private func loadBoards() async {
        let result = await boardRepository.getBoards()

        switch result {
        case .success(let boards):
            uiState.boardOptions = boards.map { BoardOptionModel(id: $0.id, name: $0.title) }
            // Set default board if none selected (prefer initialBoardId, then first board)
            if uiState.selectedBoardId == nil {
                if let initialId = initialBoardId, boards.contains(where: { $0.id == initialId }) {
                    uiState.selectedBoardId = initialId
                } else {
                    uiState.selectedBoardId = uiState.boardOptions.first?.id
                }
            }
        case .failure(let error):
            print("Failed to load boards: \(error)")
            uiState.submitError = "게시판 목록을 불러오지 못했습니다."
        }
    }

    private func buildPollInfo() -> PollCreateInfo? {
        guard uiState.isVoteEnabled else { return nil }

        let title = uiState.voteTitle.trimmingCharacters(in: .whitespacesAndNewlines)
        let options = uiState.voteOptions
            .map { $0.text.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        guard !title.isEmpty && options.count >= 2 else { return nil }

        return PollCreateInfo(title: title, options: options)
    }

    private func uploadImage(_ image: UIImage) async -> Result<String, Error> {
        // Compress and resize image
        guard let imageData = prepareImageData(image) else {
            return .failure(NSError(domain: "BoardWrite", code: -1, userInfo: [NSLocalizedDescriptionKey: "이미지 처리에 실패했습니다."]))
        }

        return await uploadRepository.uploadImage(image: imageData)
    }

    private func prepareImageData(_ image: UIImage) -> Data? {
        let maxDimension: CGFloat = 1280
        let maxBytes = 1_000_000 // ~1MB

        // Scale down if needed
        var processedImage = image
        let largestSide = max(image.size.width, image.size.height)
        if largestSide > maxDimension {
            let scale = maxDimension / largestSide
            let newSize = CGSize(
                width: image.size.width * scale,
                height: image.size.height * scale
            )
            UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
            image.draw(in: CGRect(origin: .zero, size: newSize))
            if let scaledImage = UIGraphicsGetImageFromCurrentImageContext() {
                processedImage = scaledImage
            }
            UIGraphicsEndImageContext()
        }

        // Compress to fit size limit
        var quality: CGFloat = 0.9
        var imageData = processedImage.jpegData(compressionQuality: quality)

        while let data = imageData, data.count > maxBytes, quality > 0.3 {
            quality -= 0.1
            imageData = processedImage.jpegData(compressionQuality: quality)
        }

        return imageData
    }
}
